pkg/tfgen/generate.go

```diff
@@ -322,50 +322,72 @@ func (g *Generator) makePropertyType(typePath paths.TypePath,
 		return t
 	}
 
+	// We should carry across any of the deprecation messages, to Pulumi, as per Terraform schema
+	if sch.Deprecated() != "" && elemInfo != nil {
+		elemInfo.DeprecationMessage = sch.Deprecated()
+	}
+
+	// Perform case analysis on Elem() and Type(). See shim.Schema.Elem() doc for reference. Start with scalars.
 	switch sch.Type() {
 	case shim.TypeBool:
 		t.kind = kindBool
+		return t
 	case shim.TypeInt:
 		t.kind = kindInt
+		return t
 	case shim.TypeFloat:
 		t.kind = kindFloat
+		return t
 	case shim.TypeString:
 		t.kind = kindString
-	case shim.TypeList:
-		t.kind = kindList
-	case shim.TypeMap:
-		t.kind = kindMap
-	case shim.TypeSet:
-		t.kind = kindSet
+		return t
 	}
 
-	// We should carry across any of the deprecation messages, to Pulumi, as per Terraform schema
-	if sch.Deprecated() != "" && elemInfo != nil {
-		elemInfo.DeprecationMessage = sch.Deprecated()
+	// Handle single-nested blocks next.
+	if blockType, ok := sch.Elem().(shim.Resource); ok && sch.Elem() == shim.TypeMap {
```

Comment:
Blip, this should say sch.Type() == shim.TypeMap, AND the test didn't caught this, so we need a test to cover single-nested blocks.